feat: implement chainlink oracle

andreivladbrg opened the pull request:
Closes #66 

#### Changelog

- Set optimizers to 100M
- Use public RPCs for all chains expect for mainnet and eth sepolia
    - the reason is to not have to mantain the rpcs in secrets
- Change the fee stored internally to Chainlink's 8-decimal format
- Add Chainlink price feed state variable in both factory and campaigns
- Use Chainlink to retrieve the price to calculate the wei value
- Add `setMinimumFeeToZero` in campaigns
- Add `MAX_MINIMUM_FEE` constant to 100 $
- Add `tests` dir in `src`
    - Add `ChainlinkPriceFeedAddresses` to store the price feed address that will be used during deployments
    - Add `ChainlinkPriceFeedMock` file
    - Add `InitialMinimumFees`
 - Introduce multichain fork tests
 - Add tests campaign tests for `calculateMinimumFeeInWei` and `setMinimumFeeToZero`

<details><summary>How I've retrieved the block numbers at the same timestamp</summary>
<p>

```python
import requests
import time

# Define the timestamp for February 1, 2025
timestamp = 1_738_368_000

# Define the API endpoints and parameters for each network
apis = {
    "Ethereum Mainnet": {
        "url": "https://api.etherscan.io/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourEtherscanAPIKey"
        }
    },
    "BNB Smart Chain": {
        "url": "https://api.bscscan.com/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourBscScanAPIKey"
        }
    },
    "Polygon": {
        "url": "https://api.polygonscan.com/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourPolygonScanAPIKey"
        }
    },
    "Arbitrum One": {
        "url": "https://api.arbiscan.io/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourArbiscanAPIKey"
        }
    },
    "Avalanche": {
        "url": "https://api.snowtrace.io/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourSnowtraceAPIKey"
        }
    },
    "Gnosis Chain": {
        "url": "https://api.gnosisscan.io/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourGnosisScanAPIKey"
        }
    },
    "Optimism": {
        "url": "https://api-optimistic.etherscan.io/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourOptimismScanAPIKey"
        }
    },
    "Base": {
        "url": "https://api.basescan.org/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourBaseScanAPIKey"
        }
    },
    "Linea": {
        "url": "https://api.lineascan.build/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourLineaScanAPIKey"
        }
    },
    "Scroll": {
        "url": "https://api.scrollscan.com/api",
        "params": {
            "module": "block",
            "action": "getblocknobytime",
            "timestamp": timestamp,
            "closest": "before",
            "apikey": "YourScrollScanAPIKey"
        }
    },
}

# Function to query the API and retrieve the block number
def get_block_number(api_info):
    try:
        response = requests.get(api_info["url"], params=api_info["params"])
        data = response.json()
        if data["status"] == "1":
            return data["result"]
        else:
            print(f"Error: {data['message']}")
            return None
    except Exception as e:
        print(f"Exception occurred: {e}")
        return None

# Iterate over each network and retrieve the block number
for network, api_info in apis.items():
    print(f"Querying {network}...")
    block_number = get_block_number(api_info)
    if block_number:
        print(f"Block number at timestamp {timestamp} on {network}: {block_number}")
    else:
        print(f"Could not retrieve block number for {network}.")
    time.sleep(1)  # To avoid hitting rate limits
```

</p>
</details>

PaulRBerg reviewed (changes requested):
We don't need the `get_token_prices.py` script because, as @smol-ninja pointed out [here](https://github.com/sablier-labs/reusable-workflows/pull/17#issuecomment-2678588321), we don't need to test the values returned by Chainlink.

1. Sablier only needs to respond to all possible behaviors of the Chainlink function [`latestAnswer`](https://docs.chain.link/chainlink-local/api-reference/v0.2.2/mock-v3-aggregator/#latestanswer) — which I suppose are only two: (i) return value or (ii) revert.
2. By querying the prices from an external source and checking the Chainlink values against them, **we are not testing Sablier, but Chainlink's price accuracy**. However, this is not our job - it is the job of Chainlink's employees.
3. We don't even need a special test when the returned value is 0 — it's plausible that a token may have a zero market price (e.g., when it's non-transferable). However, having a fuzz test that allows the domain to start at zero should be fine.

Further notes:

- The rationale behind not checking the values is the same as when we decided not to test that the `msg.value` leads to a higher ETH balance in `Lockup`. That's something that has to be done by Ethereum clients and Solidity core devs, not by smart contract devs.
- The free CoinGecko API is wobbly. I don't recommend using it for anything in our Solidity tests.
- Should there be a need for calling an external API, let's first try to use Bash and see if that's enough. Adding a dependency like Python increases the long-term maintenance cost significantly because [Python is a PITA to configure](https://x.com/steren/status/1890170930980094382).

andreivladbrg commented:
> Sablier only needs to respond to all possible behaviors of the Chainlink function [latestAnswer](https://docs.chain.link/chainlink-local/api-reference/v0.2.2/mock-v3-aggregator/#latestanswer) — which I suppose are only two: (i) return value or (ii) revert.
By querying the prices from an external source and checking the Chainlink values against them, we are not testing Sablier, but Chainlink's price accuracy. However, this is not our job - it is the job of Chainlink's employees.

I understand this idea, and the fact that Chainlink employees' duty is to test it thoroughly is right. Although I want to mention that we have always taken an exhaustive approach to testing our protocol, and as a precautionary method, I think we should also test it ourselves. We can never trust that there is never a mistake made.
Moreover, the current change connects our protocol to an external system, which performs a sensitive operation—providing real-time prices. 

Even more, the actual tests are not even large files ([less than 100 lines of code](https://github.com/sablier-labs/airdrops/blob/feat/chainlink-oracle/tests/fork/price-feed/ChainlinkPriceFeed.t.sol)). These tests can also be viewed as a way to test the [configured addresses](https://github.com/sablier-labs/airdrops/blob/feat/chainlink-oracle/src/tests/ChainlinkPriceFeedAddresses.sol) that will be used during factory deployments (which I think is a good thing to test).

> We don't even need a special test when the returned value is 0 — it's plausible that a token may have a zero market price (e.g., when it's non-transferable). However, having a fuzz test that allows the domain to start at zero should be fine.

I didn't quite understand this point.  

> The rationale behind not checking the values is the same as when we decided not to test that the msg.value leads to a higher ETH balance in Lockup. That's something that has to be done by Ethereum clients and Solidity core devs, not by smart contract devs.

I do believe that this is not a fair comparison. If the whole EVM implementation is wrong, then we have bigger problems. Here, we are just integrating another protocol into our system (_Interations_ from CEI), which follows the same EVM.

> The free CoinGecko API is wobbly. I don't recommend using it for anything in our Solidity tests.

I agree that the CoinGecko API is not the best, but for this specific use case, it is absolutely fine as we simply query the price of 5 tokens. CoinGecko becomes problematic when a lot of calls are being made, since it is free. (haven't faced any issue locally so far and CI seems also fine)

> Should there be a need for calling an external API, let's first try to use Bash and see if that's enough. Adding a dependency like Python increases the long-term maintenance cost significantly because [Python is a PITA to configure](https://x.com/steren/status/1890170930980094382)

I was considering writing a Bash script, and it is definitely possible to achieve the same thing, but I didn't do this for these reasons: (i) it would have been a much longer script (ii) harder to understand (iii) Python is as friendly as it gets.  

Regarding it being hard to configure, actually, it's easy IMO for any machine you use. I also intentionally didn't use external libs (like [requests](https://pypi.org/project/requests/)), especially to rely only on the standard functionalities that the language comes with (e.g., `http.client`) - to avoid having to install deps with a package manager like `pip`, as it might require some of the things mentioned in the tweet you shared. Also, given that this is a fork test, which is ignored 99% of the time during [local test runs](https://github.com/sablier-labs/airdrops/blob/9625a832233c966f7bc1c359f65efcbbce00246a/package.json#L79), this is not a PITA. Even the action step of installing it [is simple](https://github.com/sablier-labs/reusable-workflows/blob/6eba24d9ae41b6cf22650c1fa474541d83ddccf6/.github/workflows/forge-test.yml#L87-L91).

(talking of PITA to configure—have you ever tried Bash scripts on Windows? 😅)  

---


I will also let @smol-ninja share his point of view. If you both consider that this is really not worth it, I am ok with removing them. It was a nice exercise to extend the testing suite to multiple languages.

smol-ninja commented:
Even though, I personally like the idea of verifying the accuracy of Chainlink prices in the CI, I am slightly inclined towards @PaulRBerg's recommendation due to the following reasons:

Since a price API is also implemented in the UI, in the rare event when Chainlink provides price higher than the API in the UI, the UI will detect it, because the tx will fail. If the Chainlink provides price less than the API, thats OK since it represents the minimum fee.

I also think the its harder to test Chainlink's price against Coingecko. Coingecko's price is real time due to its integration with exchanges' order books. On the other hand, Chainlink's price are updated by Chainlink's partners from time to time (not real time). And therefore, its very unlikely that the two prices will have an exact match. Even if we implement a comparison with some deviation, the prices variation can still exceed the deviation in the time of high volatility, such as yesterday. So, I am not sure what can will achieve with these tests.

The worst that can happen when calling `latestAnswer` is that it will return an outdated price (updated a week back lets say) in case CL's partners don't update it. It can never revert and the contract is non-upgradeable. So, IMO, we don't even need to put it under try...check but +1 for it since its an external call.

andreivladbrg commented:
> Even though, I personally like the idea of verifying the accuracy of Chainlink prices in the CI,

I think I am being misunderstood with these tests. It is not about the **accuracy**, but more that the flow works as expected: we have an external system providing a real-time price for a token, and then we calculate a wei amount based on that token price. If the token price is 2% up or down, that's Chainlink's duty to have accurate prices. What I want to achieve is the fact that there is an actual price provided. Lmk if I made myself clear.

> I also think the its harder to test Chainlink's price against Coingecko. Coingecko's price is real time due to its integration with exchanges' order books. On the other hand, Chainlink's price are updated by Chainlink's partners from time to time (not real time). And therefore, its very unlikely that the two prices will have an exact match. Even if we implement a comparison with some deviation, the prices variation can still exceed the deviation in the time of high volatility, such as yesterday. So, I am not sure what we will achieve with these tests.

As said above, it's not about the accuracy.

> The worst that can happen when calling latestAnswer is that it will return an outdated price (updated a week back lets say) in case CL's partners don't update it. 

Agree, but this is what I want to test—the fact that there is a price (outdated or not) returned.

Also, what do you say about this point:

>>Even more, the actual tests are not even large files ([less than 100 lines of code](https://github.com/sablier-labs/airdrops/blob/feat/chainlink-oracle/tests/fork/price-feed/ChainlinkPriceFeed.t.sol)). These tests can also be viewed as a way to test the [configured addresses](https://github.com/sablier-labs/airdrops/blob/feat/chainlink-oracle/src/tests/ChainlinkPriceFeedAddresses.sol) that will be used during factory deployments (which I think is a good thing to test).

smol-ninja commented:
> but more that the flow works as expected

During the fork tests, the oracle state will be forked too which will have a value for ETHUSD. Doesn't that also test that the flow works as expected? IIUC, you need the latest value to test the whole system. Right?

> What I want to achieve is the fact that there is an actual price provided

The value will be a price at the forked timestamp.

> Even more, the actual tests are not even large files ([less than 100 lines of code](https://github.com/sablier-labs/airdrops/blob/feat/chainlink-oracle/tests/fork/price-feed/ChainlinkPriceFeed.t.sol)). These tests can also be viewed as a way to test the [configured addresses](https://github.com/sablier-labs/airdrops/blob/feat/chainlink-oracle/src/tests/ChainlinkPriceFeedAddresses.sol) that will be used during factory deployments (which I think is a good thing to test)

Nothing against this. Its nice to have if the goal is to document all chainlink addresses.

smol-ninja commented on tests/integration/concrete/campaign/shared/calculate-minimum-fee-in-wei/calculateMinimumFeeInWei.t.sol:
```diff
@@ -0,0 +1,43 @@
+        _campaignType = campaignType;
+    }
+
+    function test_GivenPriceFeedAddressZero() external {
```
Since all campaigns inherit merkle base, we can only test on one parent of merkle base (lets say merkleinstant) and assume it will work on all all the parents. Since the code is shared, no need to test it for all of them. Wdyt?

andreivladbrg replied:
> Since all campaigns inherit merkle base, we can only test on one parent of merkle base (lets say merkleinstant) and assume it will work on all all the parents. Since the code is shared, no need to test it for all of them. Wdyt

that applies to all tests from `shared` dir. so i'd say not to do this

smol-ninja replied:
OK fair point.

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -67,19 +71,25 @@ abstract contract SablierMerkleBase is
     //////////////////////////////////////////////////////////////////////////*/
 
+    /// @inheritdoc ISablierMerkleBase
+    function calculateMinimumFeeInWei() external view returns (uint256) {
```
Missing `override`

smol-ninja commented on src/tests/ChainlinkPriceFeedAddresses.sol:
```diff
@@ -0,0 +1,38 @@
+    /// @dev Populate the price feeds for the supported chains.
+    constructor() {
+        // Ethereum Mainnet
+        _priceFeeds[1] = 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419;
```
Since these values be mostly used at the time of deployment, why not put it in the Base script of this repo?

Reason behind putting them in `src/tests`?

Also, we do not need constructor since there is only one function, we can make it a library and return values in `getPriceFeedAddress`.

andreivladbrg replied:
> Since these values be mostly used at the time of deployment, why not put it in the Base script of this repo?
> Reason behind putting them in src/tests?

the reason we added `tests` dir in `evm-utils` reusability in case it is needed to be imported

> Also, we do not need constructor since there is only one function, we can make it a library and return values in getPriceFeedAddress

it would require to have many if blocks: `if(chainId == 1) return 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419` and so on
let's keep the current version

smol-ninja replied:
> the reason we added tests dir in evm-utils reusability in case it is needed to be imported

How does the same argument applies here, given that this will only be used in this repo?

> it would require to have many if blocks

Yeah but I think its suitable to be called a library than a contract in this context. The number of lines will be much less as well, you can use a single line for `if(chainId == 1) return`

andreivladbrg replied:
> How does the same argument applies here, given that this will only be used in this repo?

To avoid including the `script` dir in the package. Maybe one would like to retrieve the price feeds configured by Sablier from somewhere.

The problem I see with libraries is that you have to constantly use them like this: `MyLib.foo()`, whereas for contracts, we only need to inherit them.
You might say that this call will be made only once, but I am speaking more as a general rule

smol-ninja replied:
Why do we need to add this in package? These are constructor values which will be part of deployment broadcasts. They can retrieve the constructor args from the deployments similar to other values. Having said that, why should we treat it differently than other constructor args?

Agree with other point.

andreivladbrg replied:
To make an integration or for testing purposes—I’m not sure what they would be used for, I’m just suggesting to provide them.

> why should we treat it differently than other constructor args?

We are not, since they can be obtained from the `InitialMinimumFee` contract (from the same dir) and the admin addresses from evm-utils

smol-ninja commented on src/tests/InitialMinimumFee.sol:
```diff
@@ -0,0 +1,42 @@
+    /// @dev Populate the minimum fee mapping for the supported chains.
+    constructor() {
+        // Ethereum Mainnet
+        _minimumFees[1] = ONE_DOLLAR;
```
Reason why we need it? Since all values are `ONE_DOLLAR`, why not add a constant in utils?

andreivladbrg replied:
which utils?

> Reason why we need it?

to pass this minimum fee in deployment scripts

andreivladbrg replied:
opened this: https://github.com/sablier-labs/lockup/discussions/1183

which would include this change as well

smol-ninja replied:
Until we make a decision in that discussion, IMO we should move `InitialMinimumFee` and `ChainlinkPriceFeedAddresses`. Its really just one variable `ONE_DOLLAR` which is 1 line worth of code for which you have written 35 lines.

andreivladbrg replied:
fair enough, my idea was that the current version would be useful in case we want to set different minimum fees across chains

andreivladbrg replied:
6969625940b2c55a12f882c48216a53e3a2d4f92

smol-ninja reviewed (changes requested):
Had a brief look at `src` today. Comment below:

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,48 @@ abstract contract SablierMerkleBase is
+        }
+
+        //  Retrieve the latest price from the Chainlink price feed.
+        (, int256 price,,,) = AggregatorV3Interface(CHAINLINK_PRICE_FEED).latestRoundData();
```
Though very unlikely to revert, lets put it under try catch so that in case it reverts, it does not block `claim`.

andreivladbrg replied:
We are already performing a check in the `_setChainlinkPriceFeed` function that it does not revert, so no malicious contract can be set up. Thus, it is safe not to use `try-catch`

smol-ninja replied:
I agree but the recommendation to use try catch was as a safety mechanism to not block claim just in case. Similar to asserts that we use.

I also think it will be raised in audit as well. I saw other audit reports on code4rena relevant to oracle integration and a sanity check like this was recommended at many places.

andreivladbrg replied:
> to use try catch was as a safety mechanism to not block claim just in case. 

My problem is that I don’t see how it could revert if we already know that the address set up in the Merkle campaigns is a Chainlink contract (since we’ve already made this check in the factory).

>  also think it will be raised in audit as well. I saw other audit reports on code4rena relevant to oracle integration and a sanity check like this was recommended at many places

Maybe our design is different, as we perform this check in the factory. In the campaigns, we can be sure that the oracle is a valid contract address

smol-ninja replied:
OK fair point. In case it gets flagged in audit, then we can look into it.

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,48 @@ abstract contract SablierMerkleBase is
+        (, int256 price,,,) = AggregatorV3Interface(CHAINLINK_PRICE_FEED).latestRoundData();
+
+        // Calculate the minimum fee in wei.
+        return 1e18 * minimumFee / uint256(price);
```
Since non-USD pairs have 18 decimals ([source](https://ethereum.stackexchange.com/questions/92508/do-all-chainlink-feeds-return-prices-with-8-decimals-of-precision)), to be double sure, lets get `decimals` from `CHAINLINK_PRICE_FEED` and adjust accordingly. There is no guarantee that chainlink wont move away from 8 decimals format.

```suggestion
        return minimumFee / (uint256(price) * 10 ** (18 - oracle.decimals()));
```

assuming `minimumFee` is 18 decimals.

andreivladbrg replied:
> There is no guarantee that chainlink wont move away from 8 decimals format.

I actually thought about this, and that’s a good point. However, the reason I chose the current version is that the existing price feed contracts [which we will be using](https://github.com/sablier-labs/airdrops/blob/5040b59a00e13dd624ac8c6ff40da455932083fc/src/tests/ChainlinkPriceFeedAddresses.sol#L10-L33) have 8 decimals. They are immutable, so they cannot be changed, and making an additional external call would increase gas costs.

smol-ninja replied:
As discussed on Slack, we can use 8 decimals and since we have `MAX_FEE` implemented, there is cap on how much we can charge. So gtg here.

I will still suggest to rely on decimals because I couldn't find it in the docs of Chainlink where they have specifically mentioned that all USD pairs have 8 decimals. LMK if you have found it in the docs.

andreivladbrg replied:
> I will still suggest to rely on decimals function because I couldn't find it in the docs of Chainlink where they have specifically mentioned that all USD pairs have 8 decimals

ok, agree with this

smol-ninja reviewed (changes requested):
Some more.

smol-ninja commented on src/SablierMerkleFactoryInstant.sol:
```diff
@@ -15,12 +15,14 @@ contract SablierMerkleFactoryInstant is ISablierMerkleFactoryInstant, SablierMer
     /// @param initialMinimumFee The initial minimum fee charged for claiming an airdrop.
     constructor(
         address initialAdmin,
+        address initialChainlinkPriceFeed,
```
I see. I would recommend using `oracle` instead of using the name of Chainlink directly. I looked at other protocols and the standard is to use generic name instead of a specific protocol name.

andreivladbrg replied:
Specifically using "Chainlink" is a good thing, IMO as it makes it clear that we don’t use other oracle protocols (like Pyth). But if you have a strong preference, ok, I don’t mind

andreivladbrg replied:
f3a3fc085821b92c4661659bdbd051a4cbbf229b

andreivladbrg commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -27,10 +35,19 @@ abstract contract SablierMerkleFactoryBase is
+        Adminable(initialAdmin)
+    {
         minimumFee = initialMinimumFee;
+        _setChainlinkPriceFeed(initialChainlinkPriceFeed, initialAdmin);
```
Then it:
1. Wouldn’t emit the event.
2. In case the address is not zero, the `address(0)` check is going to be made twice.

I believe there is no winning candidate, so we can have either option

andreivladbrg replied:
f3a3fc085821b92c4661659bdbd051a4cbbf229b

smol-ninja replied:
> Wouldn’t emit the event.

Why are we not emitting oracle address in `CreateMerkleLL` events similar to fee? I think we should and in that case, we can add this check.

andreivladbrg replied:
Good idea, let's add it in the event

andreivladbrg replied:
998d6fa1626e291f89fb7adb56f8be8e001405a9

andreivladbrg commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -85,11 +112,40 @@ abstract contract SablierMerkleFactoryBase is
+            (, int256 price,,,) = AggregatorV3Interface(newChainlinkPriceFeed).latestRoundData();
+
+            // Check: the price is not zero.
+            if (price == 0) {
```
Checking the price was intended to prevent issues in case of a malicious contract. However, if the contract is malicious, it can still return a value greater than 0. So, I agree—we can simply call the function without checking the price.

andreivladbrg replied:
f3a3fc085821b92c4661659bdbd051a4cbbf229b

andreivladbrg commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -85,11 +112,40 @@ abstract contract SablierMerkleFactoryBase is
+        chainlinkPriceFeed = newChainlinkPriceFeed;
+
+        // Log the update.
+        emit SetChainlinkPriceFeed({ admin: admin, chainlinkPriceFeed: newChainlinkPriceFeed });
```
f3a3fc085821b92c4661659bdbd051a4cbbf229b

smol-ninja commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -17,6 +19,12 @@ abstract contract SablierMerkleFactoryBase is
     //////////////////////////////////////////////////////////////////////////*/
 
+    /// @inheritdoc ISablierMerkleFactoryBase
+    uint256 public constant override MAX_MINIMUM_FEE = 100e8;
```
As decided since we are going to stay with CL for atleast a year, lets go with 8 decimals to minimize arithemtic operations.

However, lets be very explicit in docs and comments (I have not reviewed comments yet) to prevent any human error.

andreivladbrg replied:
agree that we need to be very explicit in docs.

> prevent any human error.

yess, that's why it is a good idea to have `MAX_MINIMUM_FEE`

smol-ninja commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -77,6 +99,11 @@ abstract contract SablierMerkleFactoryBase is
         }
 
+        // Check: the new  fee is not greater than the maximum.
+        if (newFee > MAX_MINIMUM_FEE) {
```
Yeah it makes sense to keep `MAX_MINIMUM_FEE` in this case where we have 8 decimals so we can prevent human error. 

I agree with you on this.

smol-ninja commented on src/interfaces/ISablierMerkleBase.sol:
```diff
@@ -54,6 +62,10 @@ interface ISablierMerkleBase is IAdminable {
     /// @notice The content identifier for indexing the campaign on IPFS.
     function ipfsCID() external view returns (string memory);
 
+    /// @notice Retrieves the minimum fee required to claim the airdrop, paid in the native token of the chain.
```
Its not paid in native token anymore.

andreivladbrg replied:
yes, it is paid in the native token

what is not: the value stored in state is not in native tokens decimals

smol-ninja replied:
`Retrieves the minimum fee, in USD (8 decimals), required to claim the airdrop, to be paid in the native token of the chain`

Wouldn't this be more apt?

smol-ninja commented on src/interfaces/ISablierMerkleBase.sol:
```diff
@@ -29,14 +32,19 @@ interface ISablierMerkleBase is IAdminable {
+    /// @notice Calculates the minimum fee in wei required to claim the airdrop.
+    /// @dev It uses the `MINIMUM_FEE` and the oracle price to calculate the fee in wei.
+    /// @return The minimum fee required to claim the airdrop, as an 18-decimal number, where 1e18 is 1 native token.
+    function calculateMinimumFeeInWei() external view returns (uint256);
```
Why not just call it `minimumFeeInWei`? I understand calculations happen internally but calculations are a part of almost all the functions such as `getFee`.

andreivladbrg replied:
i considered extra explicitness good here, but no strong opinion

smol-ninja replied:
My argument was based on the fact that calculations are implicit in every functions. So, there is no need to use the word calculate.

smol-ninja replied:
I was actually comparing it with `getFee`, where we have a ternary condition inside. And for example `refundableAmountOf` in lockup also does some calculations inside but we refrain from using the word `calculate` in the function name.

smol-ninja commented on src/interfaces/ISablierMerkleBase.sol:
```diff
@@ -14,6 +14,9 @@ interface ISablierMerkleBase is IAdminable {
     /// @notice Emitted when the admin claws back the unclaimed tokens.
     event Clawback(address indexed admin, address indexed to, uint128 amount);
 
+    /// @notice Emitted when the minimum fee is set to a lower value.
```
`/// @notice Emitted when the minimum fee is reduced.`

smol-ninja commented on src/interfaces/ISablierMerkleBase.sol:
```diff
@@ -14,6 +14,9 @@ interface ISablierMerkleBase is IAdminable {
     event Clawback(address indexed admin, address indexed to, uint128 amount);
 
+    /// @notice Emitted when the minimum fee is set to a lower value.
+    event LowerMinimumFee(address indexed factoryAdmin, uint256 newFee, uint256 previousFee);
```
I'd suggest `updateMinimumFee`. `LowerMinimumFee` sounds unclear. Lowering the minimum fee is the design implementation but from users point of view, we are updating fee for him.

andreivladbrg replied:
I believe explicitly say that fee has only lowered is a good thing. 

also, we usually name the events based on the function name, so i'd leave it like this

smol-ninja commented on src/interfaces/ISablierMerkleBase.sol:
```diff
@@ -97,4 +109,13 @@ interface ISablierMerkleBase is IAdminable {
     /// @return feeAmount The amount of native tokens (e.g., ETH) collected as fees.
     function collectFees(address factoryAdmin) external returns (uint256 feeAmount);
+
+    /// @notice Sets the minimum fee to a lower value.
```
`/// @notice Sets the minimum fee to a new value.`

That new fee should be less than `minimumFee` is a part of the requirement.

smol-ninja replied:
I agree its a lower value but its still provided as the input of function. So how about?

`Sets the minimum fee to a new value lower than the current fee`

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,68 @@ abstract contract SablierMerkleBase is
+            revert Errors.SablierMerkleBase_CallerNotFactoryAdmin(factoryAdmin, msg.sender);
+        }
+
+        uint256 previousFee = minimumFee;
```
shouldn't we call it `currentFee` since at this moment its literally the current fee.

andreivladbrg replied:
sure, no preference

smol-ninja replied:
Another point is that in the next line, you used current fee in the comment.

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,68 @@ abstract contract SablierMerkleBase is
+            revert Errors.SablierMerkleBase_NewFeeNotLower(previousFee, newFee);
+        }
+
+        // Effect: set the minimum fee to zero.
```
`update the minimum fee to new value`

andreivladbrg replied:
good catch

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,68 @@ abstract contract SablierMerkleBase is
                             INTERNAL CONSTANT FUNCTIONS
     //////////////////////////////////////////////////////////////////////////*/
 
+    /// @dev See the documentation for the user-facing functions that call this internal function.
```
IMO a dev natspec would be very useful here since this returns 18 decimals.

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,68 @@ abstract contract SablierMerkleBase is
+        uint256 oracleDecimals = AggregatorV3Interface(ORACLE).decimals();
+
+        // Scale the minimum fee to 18 decimals.
+        uint256 scaledMinimumFee = minimumFee * 1e18;
```
To scale it to 18 decimals, should it not be `minimumFee * 1e10`, since `minimumFee` is already in 8 decimals?

It took time for me to understood what you are doing here. So here is my suggestion:

```solidity
        uint256 adjustedPrice;

        if (oracleDecimals == 8) {
            adjustedPrice = uint256(price);
        }
        // If the decimals is not equal to 8, adjust the price to match 8 decimals format and return.
        else if (oracleDecimals > 8) {
            adjustedPrice = uint256(price) / 10 ** (oracleDecimals - 8);
        } else {
            adjustedPrice = uint256(price) * 10 ** (oracleDecimals - 8);
        }

        // Multiply it with 1e18 before returning because the price is of for 1e18 wei tokens.
        return (minimumFee * 1e18) / adjustedPrice;
```

This is also more efficient because first check is with == 8

andreivladbrg replied:
agree, good recommendation

smol-ninja commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,68 @@ abstract contract SablierMerkleBase is
+        }
+
+        //  Retrieve the latest price from the oracle.
+        (, int256 price,,,) = AggregatorV3Interface(ORACLE).latestRoundData();
```
Since `price` is `int256` format, it can be negative. In that case, we should return 0.

```solidity
if (price < 0) { return 0; }
```

andreivladbrg replied:
can a price be actually negative? :))

but we can add this check

smol-ninja replied:
because its `int256` I don't think anybody is stopping chainlink operator for setting it negative. I dont know why they did not use `uint256` type for this. But in case that happens, claim would fail for our users.

smol-ninja reviewed (changes requested):
Comments on `src` below. For some of the feedbacks, I don't have strong opinion so will let you take the call.

andreivladbrg commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -27,10 +35,16 @@ abstract contract SablierMerkleFactoryBase is
                                     CONSTRUCTOR
     //////////////////////////////////////////////////////////////////////////*/
 
+    /// @dev Emits a {SetOracle} event if the address is not zero.
```
to not have `_setOracle` logic duplicated

smol-ninja replied:
It wont be duplicated if you move event from `_setOracle` to `setOracle` right?

andreivladbrg commented on src/interfaces/ISablierMerkleFactoryBase.sol:
```diff
@@ -30,24 +30,37 @@ interface ISablierMerkleFactoryBase is IAdminable {
-    /// @dev The fee is denominated in the native token of the chain, e.g., ETH for Ethereum Mainnet.
+    /// @dev The fee is denominated in Chainlink's 8-decimal format for USD prices, where $1 is 1e8.
     /// @param campaignCreator The address of the campaign creator.
     function getCustomFee(address campaignCreator) external view returns (MerkleFactory.CustomFee memory);
```
ser, i don't know if you remember but i made the same suggestion, and you said that you find useful to have a getter for the struct from storage (can't find where exactly) 
IMO yes, we can remove it

smol-ninja replied:
Oh. I was wrong then when I said that. 😅

Your call whether to keep it or remove it as I align with you on this now.

andreivladbrg commented on script/DeployDeterministicMerkleFactories.s.sol:
```diff
@@ -7,13 +7,16 @@ import { SablierMerkleFactoryInstant } from "../src/SablierMerkleFactoryInstant.
 /// @dev Reverts if any contract has already been deployed.
-contract DeployDeterministicMerkleFactories is BaseScript {
+contract DeployDeterministicMerkleFactories is BaseScript, ChainlinkPriceFeedAddresses {
+    uint256 private constant ONE_DOLLAR = 1e8;
```
> Also, why not move it under run() similar to other variables?

no particular preference

andreivladbrg commented on src/interfaces/ISablierMerkleFactoryBase.sol:
```diff
@@ -30,24 +30,37 @@ interface ISablierMerkleFactoryBase is IAdminable {
+    /// @notice Retrieves the maximum minimum fee that can be set.
+    /// @dev The value is equal to 100 USD.
+    /// @return The maximum minimum fee that can be set denominated in Chainlink's 8-decimal.
+    function MAX_MINIMUM_FEE() external view returns (uint256);
```
agree

andreivladbrg commented on src/interfaces/ISablierMerkleFactoryBase.sol:
```diff
@@ -30,24 +30,37 @@ interface ISablierMerkleFactoryBase is IAdminable {
+    /// @notice Retrieves the minimum fee required to claim the airdrop, paid in the native token of the
+    /// chain, e.g., ETH for Ethereum Mainnet.
+    /// @dev The fee is denominated in Chainlink's 8-decimal format for USD prices, where $1 is 1e8.
+    /// @return The minimum fee required to claim the airdrop, as an 8-decimal number.
```
fair enough

andreivladbrg commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -86,12 +105,43 @@ abstract contract SablierMerkleFactoryBase is
+
+    /// @dev See the documentation for the user-facing functions that call this internal function.
+    function _setOracle(address newOracle, address admin) private {
+        // Check: if the oracle is not zero, the call does not fail.
```
the second one looks good

andreivladbrg commented on src/libraries/Errors.sol:
```diff
@@ -27,9 +30,19 @@ library Errors {
     /// @notice Thrown when trying to claim with an invalid Merkle proof.
     error SablierMerkleBase_InvalidProof();
 
+    /// @notice Thrown when trying to set a fee that is not lower than the current fee.
```
good catch, aggre

honestly, i simply "TAB" what copilot suggest :))

smol-ninja commented on src/interfaces/ISablierMerkleFactoryBase.sol:
```diff
@@ -30,24 +30,37 @@ interface ISablierMerkleFactoryBase is IAdminable {
 
-    /// @notice Retrieves the minimum fee charged for claiming an airdrop.
-    /// @dev The fee is denominated in the native token of the chain, e.g., ETH for Ethereum Mainnet.
+    /// @notice Retrieves the minimum fee required to claim the airdrop, paid in the native token of the
```
Good point. I overlooked "paid" keyword. Lets keep it this way.

andreivladbrg commented on src/abstracts/SablierMerkleFactoryBase.sol:
```diff
@@ -86,12 +105,43 @@ abstract contract SablierMerkleFactoryBase is
     }
 
+    /// @inheritdoc ISablierMerkleFactoryBase
+    function setOracle(address newOracle) external override onlyAdmin {
```
yeah, good idea

andreivladbrg commented:
I’ve addressed your feedback in 7c90fa0b335d1562611be7bb76f930d618d596e4. I hope I haven’t missed anything.

Lmk if it looks good. Could you please mark them as "resolved" while reviewing? thanks

smol-ninja reviewed (approved):
Great work on the PR. We can merge it now for the audit. A small note that I am still reviewing the tests and will open a review PR separately. 

![pikachu](https://github.com/user-attachments/assets/b6dfbc8b-7b5d-4f15-8b11-d4a4b884cc25)

smol-ninja reviewed (commented):
Please see my review below for the tests.

I have made all the changes in https://github.com/sablier-labs/airdrops/pull/80. Only after we have agreed upon the below feedback, https://github.com/sablier-labs/airdrops/pull/80 should be merged.

andreivladbrg commented on src/abstracts/SablierMerkleBase.sol:
```diff
@@ -182,10 +195,73 @@ abstract contract SablierMerkleBase is
+    /// @dev Calculates the minimum fee in the native token with 18 decimals.
+    function _minimumFeeInWei() internal view returns (uint256) {
+        // If the oracle is not set, return 0.
+        if (ORACLE == address(0)) {
```
i wouldn’t touch `src` just for this.

both work. we can use short circuit in many more places. we’ve discussed this in the past. to avoid always debating this, why not simply _NOT_ use short circuit unless it’s absolutely necessary (e.g. [this one](https://github.com/sablier-labs/lockup/blob/7475c117f6dd6d104e94451864347140a095e108/src/abstracts/SablierLockupBase.sol#L416) - without requiring it to have nested code)? it avoids debates and makes it easier to understand when you have different `if` block checks

smol-ninja replied:
I am in soft agreement with you and can revert this change. However, we can make the following rule as well:

Use short circuit, if the returned value is same or if the revert is same. 

Wdyt about this?

> we can use short circuit in many more places

I am in favour of this as well.

smol-ninja replied:
Reverted in https://github.com/sablier-labs/airdrops/pull/80/commits/355cb2406844b09301ce25509c807f737abc3332

andreivladbrg commented on tests/integration/concrete/campaign/shared/minimum-fee-in-wei/minimumFeeInWei.t.sol:
```diff
@@ -0,0 +1,43 @@
+    string private _campaignType;
+
+    constructor(string memory campaignType) {
+        _campaignType = campaignType;
```
it wouldn’t work with the `setUp` function as it collides with `Integration_Test` setup

andreivladbrg replied:
> I am setting campaignType in Shared integration test

i believe thats not a good thing, because this variables is relevant only to `minimumFeeInWei` tests

smol-ninja replied:
Thats a good point. Let me think about it.

smol-ninja replied:
I have decided to move it back through constructor in https://github.com/sablier-labs/airdrops/pull/80/commits/8380736290e69ebb47902ddafc6cf229cee66a8d.

andreivladbrg commented on tests/integration/concrete/campaign/ll/constructor.t.sol:
```diff
@@ -18,12 +18,13 @@ contract Constructor_MerkleLL_Integration_Test is Integration_Test {
         assertEq(constructedLL.ipfsCID(), IPFS_CID, "ipfsCID");
         assertEq(address(constructedLL.LOCKUP()), address(lockup), "lockup");
         assertEq(constructedLL.MERKLE_ROOT(), MERKLE_ROOT, "merkleRoot");
+        assertEq(constructedLL.minimumFee(), MINIMUM_FEE, "minimum fee");
```
nope, we shouldn't. the values tested in the constructor are only state variables and not functions that have additional logic.  

based on your logic, we should test these here as well:  
- `getFirstClaimTime`  
- `hasClaimed`  
- `hasExpired`  

but we won't, as we have individual concrete tests for them, but for the state variables, we don't.

smol-ninja replied:
fair point.

andreivladbrg commented on tests/integration/concrete/factory/instant/constructor.t.sol:
```diff
@@ -8,13 +8,10 @@ import { Integration_Test } from "../../../Integration.t.sol";
-        assertEq(actualMinimumFee, expectedMinimumFee, "minimum fee");
+        assertEq(constructedFactoryInstant.admin(), users.admin, "factory admin");
+        assertEq(constructedFactoryInstant.oracle(), address(oracle), "oracle");
+        assertEq(constructedFactoryInstant.minimumFee(), MINIMUM_FEE, "minimum fee");
```
correct

andreivladbrg commented on tests/integration/concrete/factory/instant/constructor.t.sol:
```diff
@@ -8,13 +8,10 @@ import { Integration_Test } from "../../../Integration.t.sol";
-        uint256 actualMinimumFee = constructedFactoryInstant.minimumFee();
-        uint256 expectedMinimumFee = MINIMUM_FEE;
-        assertEq(actualMinimumFee, expectedMinimumFee, "minimum fee");
+        assertEq(constructedFactoryInstant.admin(), users.admin, "factory admin");
```
it's fine :)) since the `assertEq` statements are not formatted into multiple lines.  

if you have a strong opinion, we should make the change for all factories.

smol-ninja replied:
Yes I have made changes to all the factories.

andreivladbrg commented on tests/integration/concrete/factory/shared/reset-custom-fee/resetCustomFee.tree:
```diff
@@ -3,10 +3,10 @@ ResetCustomFee_Integration_Test
    └── when enabled
       ├── it should disable the custom fee
       ├── it should set the custom fee to 0
+      ├── it should return minimum fee
```
yeah, keeping `it should return minimum fee` is not the best

but what i'd do instead is:

```tree
├── when not enabled
│  ├── it should do nothing
│  └── it should emit a {ResetCustomFee} event
└── when enabled
   ├── it should disable the custom fee
   ├── it should set the custom fee to 0
   └── it should emit a {ResetCustomFee} event
```

wdyt?

andreivladbrg replied:
> There is no direct way to check any of these. The only check that can prove it indirectly is the following:


actually, there is a direct way in case the custom fee is enabled.

```solidity
// before
assertNotEq(merkleFactoryBase.getFee(users.campaignOwner); merkleFactoryBase.minimumFee(), "fee");
// call function
merkleFactoryBase.resetCustomFee(users.campaignOwner);
// after
assertEq(merkleFactoryBase.getFee(users.campaignOwner, merkleFactoryBase.minimumFee());
```

>  I could not find an explicit way to mention it in the branch. If we add that it should disable the custom fee , what would be the immediate assertion to test this?

but also, even if there isn't a direct way, we should highlight what the function does in english in the tree files, as they are meant to provide a high-level overview.

smol-ninja replied:
Agree with you. Incorporated your suggestion in https://github.com/sablier-labs/airdrops/pull/80/commits/de7e12e722e86c3b6fd5701eb8fb67a4acf97683.

Also, why not replace the following 

```solidity
// before
assertNotEq(merkleFactoryBase.getFee(users.campaignOwner); merkleFactoryBase.minimumFee(), "fee");
```

with 

```solidity
// Check that `getFee` returns the custom fee.
assertEq(merkleFactoryBase.getFee(users.campaignOwner), customFee, "custom fee enabled");
```

The second one also guarantee through assertion that custom fee is enabled.

andreivladbrg replied:
> The second one also guarantee through assertion that custom fee is enabled.

also mentioned in other comment, it doesn't if `minimumFee` equals the `customFee` - ik its not the case here, but IMO this is the best way to test it: `assertNotEq(merkleFactoryBase.getFee(users.campaignOwner); merkleFactoryBase.minimumFee(), "fee");`

smol-ninja replied:
Oh I got it now. Nice one on this. Agree and will make the change.

smol-ninja replied:
https://github.com/sablier-labs/airdrops/pull/80/commits/b8c514d2f68eb53b6f3ae45aebf5e3950b0f82ef

andreivladbrg commented on tests/integration/concrete/factory/shared/set-custom-fee/setCustomFee.t.sol:
```diff
@@ -14,52 +14,62 @@ abstract contract SetCustomFee_Integration_Test is Integration_Test {
     function test_WhenNotEnabled() external whenCallerAdmin {
+        uint256 newFee = 0;
+
+        assertNotEq(merkleFactoryBase.getFee(users.campaignOwner), newFee, "custom fee");
```
testing it against `MINIMUM_FEE` does not guarantee that custom fee is not enabled either, since the custom fee can also be equal to `minimumFee`. the only way we could test this is by reading the boolean value from state `enabled`, but we no longer have a getter for this.  

what we could do instead is add this after the `setCustomFee` call:  

```solidity
assertEq(merkleFactoryBase.getFee(users.campaignOwner), newFee, "custom fee");
assertNotEq(merkleFactoryBase.getFee(users.campaignOwner), merkleFactoryBase.minimumFee(), "custom fee");
```

smol-ninja replied:
if its equal to `newFee` then its definitely not equal to `minimumFee` since for the test `newFee` is not equal to `minimumFee`.

> testing it against MINIMUM_FEE does not guarantee that custom fee is not enabled either, since the custom fee can also be equal to minimumFee

Agree. I added the following at the beginning of test:

```solidity
// Check that custom fee is not enabled.
assertEq(merkleFactoryBase.getFee(users.campaignOwner), MINIMUM_FEE, "custom fee");
```

But your point is valid so I can remove it as well. I think what matters is that the test is being run with a value for custom fee which is not same as minimum fee. Given these are concrete tests, we dont need any priori validations imo.

smol-ninja replied:
https://github.com/sablier-labs/airdrops/pull/80/commits/8613269733a198e525e8ca11b4ccacb81abbad9e

andreivladbrg commented on tests/integration/concrete/factory/shared/set-oracle/setOracle.t.sol:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    function test_WhenNewOracleZeroAddress() external whenCallerAdmin {
+        resetPrank({ msgSender: users.admin });
```
correct

andreivladbrg commented on tests/integration/concrete/factory/shared/set-oracle/setOracle.t.sol:
```diff
@@ -0,0 +1,52 @@
+    function test_WhenNewOracleZeroAddress() external whenCallerAdmin {
+        resetPrank({ msgSender: users.admin });
+
+        assertNotEq(merkleFactoryBase.oracle(), address(0), "oracle before");
```
i find it a good pattern in tests to assert states before and after a function modifies it.  

> What matters is the final state.

agree, but how do we ensure that the final state we are checking was not already set before calling the function?

smol-ninja replied:
`emit ISablierMerkleFactoryBase.SetOracle(users.admin, address(0), address(oracle));`

This event ensures that previous value and the new value are not same. Doesn't it ensure the same thing?

andreivladbrg replied:
good point, forgot about the event.
we can remove it

andreivladbrg commented on tests/utils/ChainlinkPriceFeedMock.sol:
```diff
@@ -0,0 +1,22 @@
+}
+
+/// @notice A mock Chainlink price feed that does not implement the `latestRoundData` function.
+contract ChainlinkPriceFeedMock_Empty { }
```
agree

smol-ninja commented on script/CreateMerkleVCA.s.sol:
```diff
@@ -12,7 +12,11 @@ contract CreateMerkleVCA is BaseScript {
-        SablierMerkleFactoryVCA merkleFactory = new SablierMerkleFactoryVCA(DEFAULT_SABLIER_ADMIN, 0);
+        SablierMerkleFactoryVCA merkleFactory = new SablierMerkleFactoryVCA({
+            initialAdmin: DEFAULT_SABLIER_ADMIN,
+            initialMinimumFee: 0,
```
So we would have to make this change after the contracts are deployed? If so, I'd say we should make this change now so that once contracts are deployed, we dont have to worry about it.

andreivladbrg replied:
> So we would have to make this change after the contracts are deployed? If so, I'd say we should make this change now so that once contracts are deployed, we dont have to worry about it.

there is nothing to worry about. we would simply remove the deployment line, and we would have:  

```solidity
SablierMerkleFactoryVCA merkleFactory = SablierMerkleFactoryVCA(0xSOME_ADDRESS);
```

that simply doesn't matter which options we have, so we can have either of the options.

andreivladbrg commented on script/ChainlinkPriceFeedAddresses.sol:
if by "dry'ify" you mean that there are less lines of code that's depended on the formatting rules you are using (one could always have `{}` for `if` blocks) but i am ok with the change

also, for extra explicitness, we should add at the end: `return address(0);`

smol-ninja replied:
Yes, less line of code was the objective.

smol-ninja commented on tests/integration/concrete/factory/vca/create-merkle-vca/createMerkleVCA.t.sol:
```diff
@@ -135,16 +135,14 @@ contract CreateMerkleVCA_Integration_Test is Integration_Test {
         assertEq(address(actualVCA), expectedMerkleVCA, "MerkleVCA contract does not match computed address");
 
-        // It should create the campaign with 0 custom fee.
-        assertEq(actualVCA.MINIMUM_FEE(), 0, "custom fee");
```
No its not tested in constructor. This test is `test_GivenCustomFeeSet` and in constructor we are not testing default values.

andreivladbrg replied:
oh right, mb

andreivladbrg commented on script/DeployMerkleFactories.s.sol:
```diff
@@ -7,11 +7,15 @@ import { SablierMerkleFactoryInstant } from "../src/SablierMerkleFactoryInstant.
-contract DeployMerkleFactories is BaseScript {
+contract DeployMerkleFactories is BaseScript, ChainlinkPriceFeedAddresses {
+    /// @dev The initial minimum fee, using Chainlink's 8 decimals format, where 1e8 is $1.
+    uint256 private constant ONE_DOLLAR = 1e8;
```
i just realized i shouldn't have agreed to remove the `InitialMinimumFees` contract.  

we will be setting a minimum fee of $1 for chains where we don't have an oracle, so we will have `minimumFee == 1e8` and `minimumFeeInWei = 0`—that's not a good thing. we should either have a mapping (initial version) or use your suggestion: `if (chain.id == number) return some_value;`.

smol-ninja replied:
Yes agree. If the oracle is `address(0)` then minimum fee should be set to 0 as well. However, if we set minimum fee to `1e8` at the time of deployment, then when the oracle becomes active, admin would only have to set the oracle address, otherwise admin would have to set both oracle and minimum fee.

But I think chains without chainlink oracle wont have oracle for a long time. So, I can bring the `InitialMinimumFees ` contract/function back.

andreivladbrg replied:
we can add a `setOracleAndMinimumFee` function in factory base

smol-ninja replied:
https://github.com/sablier-labs/airdrops/pull/80/commits/9485b3259f257d754c4719296779edbd31322094

Added the following functions:
```solidity
    /// @notice Returns the initial minimum fee. If the chain does not have a Chainlink oracle, return 0.
    function initialMinimumFee() public view returns (uint256 fee) {
        if (chainlinkOracle() != address(0)) return 1e8;

        // Return 0 for chains without Chainlink oracle.
        return 0;
    }
```

Also added `BaseScript_Test`

smol-ninja replied:
Re `setOracleAndMinimumFee`, I think it would be redundant. Admin will most likely use safe and can therefore batch two txns.

andreivladbrg commented on tests/integration/concrete/campaign/shared/minimum-fee-in-wei/minimumFeeInWei.tree:
```diff
@@ -0,0 +1,8 @@
+MinimumFeeInWei_Integration_Test
+├── given oracle address zero
```
agree, that would be better since we added the `decimals` call

andreivladbrg commented on tests/integration/concrete/factory/shared/set-oracle/setOracle.tree:
```diff
@@ -0,0 +1,13 @@
+    │   ├── it should oracle
+    │   └── it should emit a {SetOracle} event
+    └── when new oracle not zero address
+        ├── when new oracle invalid
```
or `when new oracle without implementation` - matches the mock name

smol-ninja replied:
https://github.com/sablier-labs/airdrops/pull/80/commits/5761aacb933a09b9010c47cf8784e7bf98ee75c9

andreivladbrg commented on tests/fork/merkle-campaign/MerkleInstant.t.sol:
```diff
@@ -180,7 +181,7 @@ abstract contract MerkleInstant_Fork_Test is Fork_Test {
         expectCallToClaimWithData({
             merkleLockup: address(vars.merkleInstant),
-            fee: MINIMUM_FEE,
+            feeInWei: MINIMUM_FEE_IN_WEI,
```
agree

smol-ninja commented on tests/fork/price-feed/ChainlinkPriceFeed.t.sol:
```diff
@@ -0,0 +1,88 @@
+    mapping(string chainName => ForkData forkData) internal _forkData;
+
+    function setUp() public override {
+        _forkData["mainnet"] = ForkData({ blockNumber: 21_747_949, nativeTokenPrice: 3296.39063484e8 });
```
https://github.com/sablier-labs/airdrops/pull/80/commits/f2521935f4b89a48115281819c31d72b0210d611#diff-16b1f030d830c4b4daab51316825206a7cccdc054ee22a003c69a56a6e99a31b
